main/src/ca/uwaterloo/flix/language/ast/UnkindedType.scala

```diff
@@ -0,0 +1,601 @@
+/*
+ * Copyright 2021 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.ast
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+/**
+  * A type without an associated kind.
+  */
+sealed trait UnkindedType {
+
+  /**
+    * Returns the type constructor of the given type.
+    */
+  def typeConstructor: Option[UnkindedType.Constructor] = this match {
+    case UnkindedType.Cst(cst, _) => Some(cst)
+    case UnkindedType.Apply(t1, _) => t1.typeConstructor
+    case UnkindedType.Var(_, _, _) => None
+    case UnkindedType.Ascribe(t, _, _) => t.typeConstructor
+    case UnkindedType.Lambda(_, _) => throw InternalCompilerException("Unexpected type constructor: Lambda")
+  }
+
+  /**
+    * Returns the base type of a type application.
+    */
+  def baseType: UnkindedType.BaseType = this match {
+    case UnkindedType.Apply(t1, _) => t1.baseType
+    case bt: UnkindedType.BaseType => bt
+  }
+
+  /**
+    * Returns the arguments of a type application.
+    */
+  def typeArguments: List[UnkindedType] = this match {
+    case UnkindedType.Apply(tpe1, tpe2) => tpe1.typeArguments ::: tpe2 :: Nil
+    case UnkindedType.Ascribe(t, _, _) => t.typeArguments
+    case _ => Nil
+  }
+}
+
+object UnkindedType {
+
+  /**
+    * The union of non-Apply types.
+    */
+  sealed trait BaseType
+
+  /**
+    * Type constructor.
+    */
+  case class Cst(cst: Constructor, loc: SourceLocation) extends UnkindedType with BaseType
+
+  /**
+    * Type application.
+    */
+  case class Apply(t1: UnkindedType, t2: UnkindedType) extends UnkindedType
+
+  /**
+    * Type lambda.
+    */
+  case class Lambda(t1: UnkindedType.Var, t2: UnkindedType) extends UnkindedType with BaseType
+
+  /**
+    * Type variable.
+    */
+  case class Var(id: Int, text: Option[String] = None, loc: SourceLocation) extends UnkindedType with BaseType {
+    /**
+      * Converts the UnkindedType to a Type with the given `kind`.
+      */
+    def ascribedWith(kind: Kind): Type.Var = Type.Var(id, kind, text = text)
+
+    override def equals(other: Any): Boolean = other match {
+      case Var(otherId, _, _) => id == otherId
```

Comment:
I am surprised. This pattern match is not exhaustive, yet the code compiles?????????????????????????????

Reply:
Yes, very weird...